sklearn/ensemble/_hist_gradient_boosting/gradient_boosting.py

```diff
@@ -0,0 +1,758 @@
+        # time spent predicting X for gradient and hessians update
+        acc_prediction_time = 0.
+        X, y = check_X_y(X, y, dtype=[X_DTYPE])
+        y = self._encode_y(y)
+        rng = check_random_state(self.random_state)
+
+        self._validate_parameters()
+        self.n_features_ = X.shape[1]  # used for validation in predict()
+
+        # we need this stateful variable to tell raw_predict() that it was
+        # called from fit() (this current method), and that the data it has
+        # received is pre-binned.
+        # predicting is faster on pre-binned data, so we want early stopping
+        # predictions to be made on pre-binned data. Unfortunately the scorer_
+        # can only call predict() or predict_proba(), not raw_predict(), and
+        # there's no way to tell the scorer that it needs to predict binned
+        # data.
+        self._in_fit = True
+
+        # bin the data
+        if self.verbose:
+            print("Binning {:.3f} GB of data: ".format(X.nbytes / 1e9), end="",
+                  flush=True)
+        tic = time()
+        self.bin_mapper_ = _BinMapper(max_bins=self.max_bins, random_state=rng)
+        X_binned = self.bin_mapper_.fit_transform(X)
+        toc = time()
+        if self.verbose:
+            duration = toc - tic
+            print("{:.3f} s".format(duration))
+
+        self.loss_ = self._get_loss()
+
+        self.do_early_stopping_ = (self.n_iter_no_change is not None and
+                                   self.n_iter_no_change > 0)
+
+        # create validation data if needed
+        if self.do_early_stopping_ and self.validation_fraction is not None:
+            # stratify for classification
+            stratify = y if hasattr(self.loss_, 'predict_proba') else None
+
+            X_binned_train, X_binned_val, y_train, y_val = train_test_split(
+                X_binned, y, test_size=self.validation_fraction,
+                stratify=stratify, random_state=rng)
+            if X_binned_train.size == 0 or X_binned_val.size == 0:
+                raise ValueError(
+                    'Not enough data (n_samples={}) to '
+                    'perform early stopping with validation_fraction='
+                    '{}. Use more training data or '
+                    'adjust validation_fraction.'.format(
+                        X_binned.shape[0],
+                        self.validation_fraction)
+                )
+            # Predicting is faster of C-contiguous arrays, training is faster
+            # on Fortran arrays.
+            X_binned_val = np.ascontiguousarray(X_binned_val)
+            X_binned_train = np.asfortranarray(X_binned_train)
+        else:
+            X_binned_train, y_train = X_binned, y
+            X_binned_val, y_val = None, None
+
+        # Subsample the training set for early stopping and score monitoring
+        if self.do_early_stopping_:
+            subsample_size = 10000  # should we expose this parameter?
```

Comment:
What about the comment? Are there any thoughts about exposing this?
Do other algorithms which support early stopping have something similar?

Reply:
> What about the comment? Are there any thoughts about exposing this?

We decided to keep it as is, and expose it if users tell us they need it. I'll add a comment in the `train_score_` docstring though.

>Do other algorithms which support early stopping have something similar?

I don't think so.

Also, it's not that critical given the size of the subset, but we should ideally stratify (will be possible when #13549 is merged)